embedded/store/kv_metadata.go

```diff
@@ -15,43 +15,200 @@ limitations under the License.
+	return nil
+}
+
+func (a *deletedAttribute) deserialize(b []byte) (int, error) {
+	return 0, nil
+}
+
+type expiresAtAttribute struct {
+	expiresAt time.Time
 }
 
-const deletedFlag = 1 << 7
+func (a *expiresAtAttribute) code() attributeCode {
+	return expiresAtAttrCode
+}
+
+func (a *expiresAtAttribute) serialize() []byte {
+	var b [tsSize]byte
+	binary.BigEndian.PutUint64(b[:], uint64(a.expiresAt.Unix()))
+	return b[:]
+}
+
+func (a *expiresAtAttribute) deserialize(b []byte) (int, error) {
+	if len(b) < tsSize {
+		return 0, ErrIllegalArguments
+	}
+
+	a.expiresAt = time.Unix(int64(binary.BigEndian.Uint64(b)), 0)
+
+	return tsSize, nil
+}
 
 func NewKVMetadata() *KVMetadata {
-	return &KVMetadata{}
+	return &KVMetadata{
+		attributes: make(map[attributeCode]attribute),
+	}
 }
 
 func (md *KVMetadata) AsDeleted(deleted bool) *KVMetadata {
-	md.deleted = deleted
+	if !deleted {
+		delete(md.attributes, deletedAttrCode)
+		return md
+	}
+
+	_, ok := md.attributes[deletedAttrCode]
+	if !ok {
+		md.attributes[deletedAttrCode] = &deletedAttribute{}
+	}
+
 	return md
 }
 
 func (md *KVMetadata) Deleted() bool {
-	return md.deleted
+	_, ok := md.attributes[deletedAttrCode]
+	return ok
+}
+
+func (md *KVMetadata) ExpiresAt(expiresAt time.Time) *KVMetadata {
+	expAtAttr, ok := md.attributes[expiresAtAttrCode]
+	if !ok {
+		expAtAttr = &expiresAtAttribute{expiresAt: expiresAt}
+		md.attributes[expiresAtAttrCode] = expAtAttr
+		return md
+	}
+
+	expAtAttr.(*expiresAtAttribute).expiresAt = expiresAt
+	return md
+}
+
+func (md *KVMetadata) NonExpirable() *KVMetadata {
+	delete(md.attributes, expiresAtAttrCode)
+	return md
+}
+
+func (md *KVMetadata) IsExpirable() bool {
+	_, ok := md.attributes[expiresAtAttrCode]
+	return ok
+}
+
+func (md *KVMetadata) ExpirationTime() (time.Time, error) {
+	expAtAttr, ok := md.attributes[expiresAtAttrCode]
+	if !ok {
+		return time.Now(), ErrNonExpirable
+	}
+
+	return expAtAttr.(*expiresAtAttribute).expiresAt, nil
+}
+
+func (md *KVMetadata) ExpiredAt(mtime time.Time) bool {
+	expAtAttr, ok := md.attributes[expiresAtAttrCode]
+	if !ok {
+		return false
+	}
+
+	return !expAtAttr.(*expiresAtAttribute).expiresAt.After(mtime)
 }
 
 func (md *KVMetadata) Bytes() []byte {
-	var b byte
+	var b bytes.Buffer
+
+	deletedAttr, ok := md.attributes[deletedAttrCode]
+	if ok {
+		b.WriteByte(byte(deletedAttr.code()))
+		b.Write(deletedAttr.serialize())
+	}
 
-	if md.deleted {
-		b = b | deletedFlag
+	expAtAttr, ok := md.attributes[expiresAtAttrCode]
+	if ok {
+		b.WriteByte(byte(expAtAttr.code()))
+		b.Write(expAtAttr.serialize())
 	}
 
-	return []byte{b}
+	return b.Bytes()
 }
 
 func (md *KVMetadata) ReadFrom(b []byte) error {
-	if len(b) == 0 {
-		return nil
-	}
+	i := 0
+
+	for {
+		if len(b) == i {
+			break
+		}
+
+		if len(b[i:]) < attrCodeSize {
+			return ErrCorruptedData
+		}
+
+		attrCode := attributeCode(b[i])
+		i += attrCodeSize
 
-	md.deleted = b[0]&deletedFlag != 0
+		switch attrCode {
```

Comment:
Post-merge: We should check for duplicates here, maybe even require that attributes appear in increasing order of attrCode? It's to avoid confusion if an attribute is duplicated

Reply:
yes, can be added later on (tampering changing the state will be detected)